### PR TITLE
Convert blocking-elements to TypeScript

### PR DESCRIPTION
**Pull request by aomarks**
- Convert to TypeScript. NOTE: I strongly recommend reading the "Convert to TypeScript" commit (cd01c88) separately from the rest.

- [BREAKING] Remove WebComponents V0 support

- [BREAKING] `/dist/blocking-elements.js` is now ES2017 instead of ES5 (i.e. it now has classes). `/dist/blocking-elements.min.js` is still ES5. `/blocking-elements.js` doesn't exist anymore.

- [BREAKING] Remove HTML import file.

- Clang-format, and replace eslint with tslint.
- `dist/` files are no longer checked in, instead we ensure `npm run build` is run in `prepack`.
- Add full text LICENSE file.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,6 @@
+dist/**
 node_modules
-bower_components
```
Comment (valdrinkoshi) [inline .gitignore]: 💯

File: package.json [hunk 2]
```diff
@@ -2,26 +2,28 @@
   "name": "blocking-elements",
   "version": "0.0.2",
   "description": "A polyfill for the proposed blocking elments stack API",
-  "main": "dist/blocking-elements.js",
+  "main": "blocking-elements.js",
```
Comment (valdrinkoshi) [inline package.json]: should this be .ts?

File: rollem.config.js [hunk 3]
```diff
@@ -4,21 +4,7 @@ const babel = require('rollup-plugin-babel');
...
-      main: true
-    }),
-    commonjs(),
-    babel({
-      exclude: 'node_modules/**', // only transpile our source code
-    }),
-  ],
-}, {
-  entry: 'blocking-elements.js',
+  entry: 'dist/blocking-elements.js',
```
Comment (valdrinkoshi) [inline rollem.config.js]: The generated js does not have an IIFEE

File: blocking-elements.ts [hunk 4]
```diff
@@ -0,0 +1,407 @@
+/**
+ * @license
```
Comment (valdrinkoshi) [inline blocking-elements.ts]: license currently not included in the dist/* files

File: blocking-elements.ts [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+  private[_alreadyInertElements] = new Set<IntertableHTMLElement>();
+
+  /**
+   * Call this whenever this object is about to become obsolete. This empties
+   * the blocking elements
+   */
+  destructor() {
+    // Restore original inertness.
+    this[_restoreInertedSiblings](this[_topElParents]);
+    const nullable = this as unknown as {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Yep, exactly. Done.

File: blocking-elements.ts [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+   */
+  get top(): HTMLElement|null {
+    const elems = this[_blockingElements];
+    return elems[elems.length - 1] || null;
+  }
+
+  /**
+   * Adds the element to the blocking elements.
+   */
+  push(element: HTMLElement): HTMLElement|undefined {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Done.

File: blocking-elements.ts (hunk 4 above)
Comment (aomarks) [reply blocking-elements.ts]: Fixed. For some reason the rollup license-preserving text wasn't working. I upgraded to the latest rollup and babel, removed rollem (not needed anymore), and also tweaked the script to only emit each unique license once (since there are now two input files).

File: rollem.config.js (hunk 3 above)
Comment (aomarks) [reply rollem.config.js]: Fixed.

As discussed in chat, `/dist/blocking-elements.js` is now an ES module that exports a singleton instead of setting the `document.$blockingElements` global. The `.min.js` file does still set the global, and continues to be IIFE-guarded ES5 minified. Included usage instructions in the README.

File: package.json (hunk 2 above)
Comment (aomarks) [reply package.json]: Oops, this shouldn't actually have changed.

File: blocking-elements.ts [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+      for (let i = 0; i < mutation.addedNodes.length; i++) {
+        const sibling = mutation.removedNodes[i] as IntertableHTMLElement;
+        if (!this[_isInertable](sibling)) {
+          continue;
+        }
+        if (toKeepInert && sibling.inert) {
+          toKeepInert.add(sibling);
+        } else {
+          sibling.inert = true;
+          if (inertedSiblings) {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Fixed, see other comment.

File: blocking-elements.ts [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+
+      // To restore.
+      for (let i = 0; i < mutation.removedNodes.length; i++) {
+        const sibling = mutation.removedNodes[i] as IntertableHTMLElement;
+        if (sibling === inertedChild) {
+          console.info('Detected removal of the top Blocking Element.');
+          this.pop();
+          return;
+        }
+        if (inertedSiblings && inertedSiblings.has(sibling)) {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Fixed, see other comment.

File: blocking-elements.ts [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+          inertedSiblings.add(sibling);
+        }
+      }
+      // Store the siblings that were inerted.
+      element[_siblingsToRestore] = inertedSiblings;
+      // Observe only immediate children mutations on the parent.
+      element[_parentMO] =
+          new MutationObserver(this[_handleMutations].bind(this));
+      const mo = element[_parentMO];
+      if (element.parentNode !== null && mo !== undefined) {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Done.

File: blocking-elements.ts [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+   * Sets the property `inert` over the attribute since the inert spec
+   * doesn't specify if it should be reflected.
+   * https://html.spec.whatwg.org/multipage/interaction.html#inert
+   */
+  private[_inertSiblings](
+      elements: IntertableHTMLElement[], toSkip: Set<HTMLElement>|null,
+      toKeepInert: Set<HTMLElement>|null) {
+    for (const element of elements) {
+      const children =
+          element.parentNode !== null ? element.parentNode.children : [];
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Done (with comment).

File: blocking-elements.ts [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+      const inertedSiblings = new Set<HTMLElement>();
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Fixed, see other comment.

File: blocking-elements.ts [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+  /**
+   * Restores original inertness to the siblings of the elements.
+   * Sets the property `inert` over the attribute since the inert spec
+   * doesn't specify if it should be reflected.
+   * https://html.spec.whatwg.org/multipage/interaction.html#inert
+   */
+  private[_restoreInertedSiblings](elements: IntertableHTMLElement[]) {
+    elements.forEach((el) => {
+      const mo = el[_parentMO];
+      if (mo !== undefined) {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Fixed, see other comment.

File: blocking-elements.ts [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+   * https://html.spec.whatwg.org/multipage/interaction.html#inert
+   */
+  private[_swapInertedSibling](
+      oldInert: IntertableHTMLElement, newInert: IntertableHTMLElement): void {
+    const siblingsToRestore = oldInert[_siblingsToRestore];
+    // oldInert is not contained in siblings to restore, so we have to check
+    // if it's inertable and if already inert.
+    if (this[_isInertable](oldInert) && !oldInert.inert) {
+      oldInert.inert = true;
+      if (siblingsToRestore) {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply blocking-elements.ts]: Done. I minimized the amount of casting we need by splitting into `HasInternalState` and a `MaybeHasInternalState`, and only casting for `_topElParents`, which was the only place we needed it.

**Comment by aomarks** [comment]
PTAL! Note I moved the source files (of which there are now two) into `src/`, sorry for the churn.

File: rollem.config.js (hunk 3 above)
Comment (valdrinkoshi) [reply rollem.config.js]: Uhm, I'm wondering if we should rather create a new `dist/blocking-elements-module.js` for the ES module case. The README.md is centered around the `document.$blockingElements` API, so probably `dist/blocking-elements.js` and `dist/blocking-elements.min.js` should differ only on minification (both should be IFEEd and set the `document.$blockingElements` global).

File: src/blocking-elements.ts [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+const _getParents = Symbol();
+const _getDistributedChildren = Symbol();
+const _isInertable = Symbol();
+const _handleMutations = Symbol();
+
+interface Inertable extends HTMLElement {
+  inert?: boolean;
+}
+
+interface InternalState {
```
Comment (valdrinkoshi) [inline src/blocking-elements.ts]: Throwing this here: the BlockingElements class could as well keep track of all this via couple WeakMaps, but that would be quite some refactoring (which probably does belong to a separate PR)

File: src/blocking-elements.ts [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+  remove(element: HTMLElement): boolean {
+    const i = this[_blockingElements].indexOf(element);
+    if (i === -1) {
+      return false;
+    }
+    this[_blockingElements].splice(i, 1);
+    // Top changed only if the removed element was the top element.
+    if (i === this[_blockingElements].length) {
+      const top = this.top;
+      if (top !== null) {
```
Comment (valdrinkoshi) [inline src/blocking-elements.ts]: this check is not needed, _topChanged allows for `top === null`:

```js
this[_topChanged](this.top);
```

File: src/blocking-elements.ts [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+   * doesn't specify if it should be reflected.
+   * https://html.spec.whatwg.org/multipage/interaction.html#inert
+   */
+  private[_restoreInertedSiblings](elements: HasInternalState[]) {
+    for (const element of elements) {
+      const mo = element[_parentMO];
+      mo.disconnect();
+      (element as MaybeHasInternalState)[_parentMO] = undefined;
+      const siblings = element[_siblingsToRestore];
+      if (siblings !== undefined) {
```
Comment (valdrinkoshi) [inline src/blocking-elements.ts]: can remove this check

File: src/blocking-elements.ts (hunk 15 above)
Comment (aomarks) [reply src/blocking-elements.ts]: Done.

File: src/blocking-elements.ts (hunk 16 above)
Comment (aomarks) [reply src/blocking-elements.ts]: Done.

File: package.json [hunk 17]
```diff
@@ -4,43 +4,44 @@
...
-  "homepage": "https://github.com/PolymerLabs/blockingElements#readme",
+  "homepage": "https://github.com/PolymerLabs/blocking-elements#readme",
   "scripts": {
-    "lint": "eslint blocking-elements.js",
+    "lint": "tslint --project ./",
     "test": "npm run lint && npm run build && easy-sauce",
-    "build": "rollem"
+    "format": "clang-format --style=file -i src/*.ts",
+    "prepack": "npm run build",
+    "build": "rm dist/* -f && tsc && rollup -c && rm dist/document-global.*"
```
[the remark this answers is not in the document]
Comment (aomarks) [reply package.json]: Actually I wanted to `rm dist/*` instead of `rm dist` on on purpose because in the past I've been bitten by tools that are watching the directory's *inode* getting confused by the directory going away. Rewrote a little though.

File: rollem.config.js (hunk 3 above)
Comment (aomarks) [reply rollem.config.js]: Agreed, it is a little weird. Both files now set `document.$blockingElement`.

**Comment by aomarks** [comment]
PTAL!

This includes a fix for an exception that was being thrown in the tests when ShadyDOM was enabled. I'm not 100% sure about it, but it does make the tests pass.

In investigating that issue, I also found a case that crashes with native shadow roots, described at https://github.com/PolymerLabs/blocking-elements/issues/18. I included a test case for it in this PR (currently marked as `.skip`).

Also, I have not been able to get Safari to report success from Sauce, even though from the Sauce videos, Safari is passing 100% of tests. I assume this is some issue with Safari/sauce/easy-sauce. For now I have removed Safari and filed https://github.com/PolymerLabs/blocking-elements/issues/17 to investigate later. I also bumped the version of Edge we're testing from 14 to "latest".
